server/spec/refresh_pools_spec.rb

```diff
@@ -214,6 +214,154 @@
+    owner_key = random_string('test_owner')
+    owner = create_owner(owner_key)
+
+    product = create_upstream_product(random_string('test_prod'))
+    sub = create_upstream_subscription(random_string('test_sub'), owner_key, {
+      :product => product,
+      :start_date => Date.today - 20,
+      :end_date => Date.today + 20
+    })
+
+    @cp.refresh_pools(owner_key)
+    pools = @cp.list_pools({:owner => owner.id})
+    expect(pools.length).to eq(1)
+
+    user = user_client(owner, random_string('test_user'))
+    consumer = consumer_client(user, random_string('test_consumer'))
+
+    entitlements = consumer.consume_pool(pools.first['id'])
+    expect(entitlements.length).to eq(1)
+
+    consumer_entitlements = @cp.list_entitlements({:uuid => consumer.uuid})
+    expect(consumer_entitlements.length).to eq(1)
+    entitlement = consumer_entitlements.first
+
+    # Update subscription, then refresh. The entitlements should be regenerated
+    update_upstream_subscription(sub.id, { :start_date => Date.today - 10 })
+    @cp.refresh_pools(owner_key)
+
+    consumer_entitlements = @cp.list_entitlements({:uuid => consumer.uuid})
+    expect(consumer_entitlements.length).to eq(1)
+    new_entitlement = consumer_entitlements.first
+
+    expect(new_entitlement['pool']['id']).to eq(entitlement['pool']['id'])
+    expect(new_entitlement['certificates'].first['id']).to_not eq(entitlement['certificates'].first['id'])
+  end
+
+  it 'revokes entitlements when pool end date changes' do
+    owner_key = random_string('test_owner')
+    owner = create_owner(owner_key)
+
+    product = create_upstream_product(random_string('test_prod'))
+    sub = create_upstream_subscription(random_string('test_sub'), owner_key, {
+      :product => product,
+      :start_date => Date.today - 20,
+      :end_date => Date.today + 20
+    })
+
+    @cp.refresh_pools(owner_key)
+    pools = @cp.list_pools({:owner => owner.id})
+    expect(pools.length).to eq(1)
+
+    user = user_client(owner, random_string('test_user'))
+    consumer = consumer_client(user, random_string('test_consumer'))
+
+    entitlements = consumer.consume_pool(pools.first['id'])
+    expect(entitlements.length).to eq(1)
+
+    consumer_entitlements = @cp.list_entitlements({:uuid => consumer.uuid})
+    expect(consumer_entitlements.length).to eq(1)
+    entitlement = consumer_entitlements.first
+
+    # Update subscription, then refresh. The entitlements should be regenerated
+    update_upstream_subscription(sub.id, { :end_date => Date.today + 10 })
+    @cp.refresh_pools(owner_key)
+
+    consumer_entitlements = @cp.list_entitlements({:uuid => consumer.uuid})
+    new_entitlement = consumer_entitlements.first
+
+    expect(new_entitlement['pool']['id']).to eq(entitlement['pool']['id'])
+    expect(new_entitlement['certificates'].first['id']).to_not eq(entitlement['certificates'].first['id'])
+  end
+
+  it 'revokes entitlements when pool product changes' do
```

Comment:
The comment on the flush for 1567922 is not as detailed as commit summary of the fix. Specifically, it mentions a scenario where entitlement-derived pools being deleted (not just the entitlements being regenerated) after pool/product changes.

So to me it looks like the scenario that could potentially trigger a problem is: 
- We have a main pool with a finite quantity, and also a virt_limit attribute on it
- A host consumes entitlement(s) from it
- A derived pool is created as a result of that entitlement attachment
- The main pool is changed in one of two ways: 
    a) the quantity of it is reduced, resulting in the host's entitlement being revoked (because now it would be overconsuming), and that would mean we need to delete the derived pool that was created because of it, or
    b) virt_limit is removed from the main pool's product (meaning it no longer should provide derived pools),
which would mean we need to delete the derived pool
- That derived pool that was marked for deletion is attempted to be locked+updated later on, which results on some kind of error

Reply:
Ah ha. I'll add a test for this scenario so we don't have to worry about a regression here.

Reply:
Done.